doc/ADR/ADR_001.md

```diff
@@ -0,0 +1,138 @@
+
+## Context
+
+This diagram shows the initial design of the Solarkraft architecture: 
+
+![diagram](./Solarkraft_components.png) 
+
+To create this initial architecture, three core components are necessary:
+
+	1. A data representation of a smart contract
+	2. A Transition Extractor
+	3. A Monitor Executor
+
+We describe each component, and their interactions with the other two, separately below.
+	
+### Smart contracts
+Smart contracts are the base objects, which are stored, modified, or inspected by various Solarkraft components.
+
+While smart contracts in e.g. Soroban feature fully executable code, for the purposes of runtime monitoring we do not have to capture their functionality whatsoever. 
+Thus, for our purposes, smart contracts consist of 
+
+	1. Smart contract State, and
+	2. Smart contract Interface
+
+Here, _State_ is an unordered collection of _named fields_, each of which holds a value of [some type](types). Note that we are working with a significantly simplified set of types in practice;
+The types we actually have to deal with boil down to primitives, arrays (Vectors), and mappings, and some basic wrapper combinations thereof in structs/enums.
+
+Thus, _State_ is inherently JSON-like, since we already know, from the design of Apalache's ITF, how to represent maps and arrays/vectors (representing structs/enums should be straightforward as well). 
+
+In contrast, _Interface_ is simply an unordered collection of the public methods admitted by a contract, and can be thought of a simple set or array.
+
+It is important to emphasize, that we do not need to capture function behavior in any way; the Transition Extractor observes a history of transactions that have already happened, and the effects of functions are observed as changes to the state of a contract, from one block to another.
+
+### Transition Extractor
+
+This component reads transaction data from the Soroban RPC, and builds a transaction log; an ordered sequence of smart contract States (or tuples of States, if we're observing a collection of contracts), annotated with the names of the functions called at each step.
+
+From a technical perspective, a key feature of the Transition Extractor, is dealing with the way the Soroban RPC serves data - the XDR format.
+
+There are several [RPC methods](RPC) avaliable, though two are of particular interest to us:
+
+	- `getLedgerEntries`, which allows us to probe any contract state, provided we can interpret the result XDR
+	- `getTransaction`, which allows us to determine which functions were called. 
+	
+Using `getLedgerEntries`, we should be able to construct a smart contract _State_ (as defined above) object, and avoid XDR in other components of Solarkraft. [This guide](RPCguide) should be used to get a general idea of the steps required here.
+
+#### Interface
+We envision the following methods, as part of the Transition Extractor interface:
+
+```
+fn readState(contractId: ContractId) -> State
```

Comment:
> Shouldn't this also take a block number?

Yes and no. Abstractly, yes, but from what I understand of the RPC methods (https://developers.stellar.org/network/soroban-rpc/methods), they themselves only read from the current state, so I'm not sure how you'd implement the RPC calls under the hood if you wanted blocks in the past.

> What is contractId / ContractId? An address?

This is from that link you found that I haven't embedded yet (https://developers.stellar.org/docs/smart-contracts/guides/rpc/retrieve-contract-code-python). Quote: "When you deploy a contract, first the code is "installed" (i.e., it is uploaded onto the blockchain). This creates a LedgerEntry containing the Wasm byte-code, which is uniquely identified by its hash (that is, the hash of the uploaded code itself)"

Reply:
Not necessarily. I imagine an MVP could look like this: You have a monitor with a state property. You read 1 state from the rpc, and verify that this state satisfies the property.

In the next step (v1.1), you tweak the initialization to, instead of reading 1 state, read a sequence of states  (which may or may not require reading them one at a time from the rpc and fast-forwarding or rewinding it), and make your TLA spec represent `P(init) /\ P(s1) /\ ... /\ P(sn)`

RE the other link, I'll add it in.

Reply:
Yeah, actually I don't think you want to use the Soroban RPC, but [Horizon](https://developers.stellar.org/network/horizon) instead:

> For comprehensive debugging needs that extend beyond the 2-hour timeframe, it is advisable to retrieve transaction information from Horizon, as it provides a lasting and persistent record.
https://developers.stellar.org/network/soroban-rpc/methods/getTransaction